### PR TITLE
Create Kafka and KafkaConnect clusters when using strimzi provider

### PR DESCRIPTION
**Pull request by bsquizz**
The `providers.kafka` section of a `ClowdEnvironment` now has been reworked with this format:
```
kafka:
  mode: operator
  pvc: true
  cluster:
    namespace: kafka-ns
    name: my-kafka-cluster
    replicas: 1
  connect:
    namespace: connect-ns
    name: my-kafka-connect-cluster
    replicas: 1
```

* When in operator mode, Clowder now takes care of dropping a Strimzi `Kafka` and `KafkaConnect` resource into the namespaces defined in your kafka provider config. In `operator` mode, other options can be tweaked to control behavior of the provisioner as well... see the `KafkaClusterConfig` and `KafkaConnectClusterConfig` ClowdEnvironment types.
* In app-interface mode, only the `name` and `namespace` are used.
* The `connect` and `cluster` sections are entirely ignored in `local` or `none` mode.

### CODE REVIEW

**Comment by bsquizz** [comment]
[test]

File: build/e2e-test.sh [hunk 1]
```diff
@@ -3,9 +3,7 @@ set -e
 
 export IMAGE_TAG=`git rev-parse --short HEAD`
 
-kubectl apply -f build/skuttl-namespace.yaml
-kubectl apply -f build/skuttl-perms.yaml
-kubectl apply -f build/prommie-operator-bundle.yaml
+bash build/setup_minikube.sh
```
Comment (psav) [inline build/e2e-test.sh:6]: This script is incorrect - it should be `minikube_setup.sh`

File: build/pr_check_inner.sh [hunk 2]
```diff
@@ -62,16 +60,41 @@ export PATH="$KUBEBUILDER_ASSETS:$PATH"
...
+mv kuttl-test.xml artifacts/junit-kuttl.xml
+
+CLOWDER_PODS=$(kubectl get pod -n clowder-system -o jsonpath='{.items[*].metadata.name}')
+for pod in $CLOWDER_PODS; do
+    kubectl logs $pod -n clowder-system > artifacts/$pod.log
+done
+
+STRIMZI_PODS=$(kubectl get pod -n strimzi -o jsonpath='{.items[*].metadata.name}')
+for pod in $STRIMZI_PODS; do
+    kubectl logs $pod -n clowder-system > artifacts/$pod.log
```
Comment (psav) [inline build/pr_check_inner.sh:90]: The namespace here is wrong - should be `strimzi`

File: bundle/tests/scorecard/kuttl/test-cyndi-strimzi/00-install.yaml [hunk 3]
```diff
@@ -13,3 +13,28 @@ metadata:
...
+        tls: false
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 1
+    storage:
+      type: ephemeral
+  entityOperator:
+    topicOperator: {}
+    userOperator: {}
```
Comment (psav) [inline bundle/tests/scorecard/kuttl/test-cyndi-strimzi/00-install.yaml:40]: New lines at end of file please :)

File: bundle/tests/scorecard/kuttl/test-kafka-strimzi-kafka-provisioner/00-install.yaml [hunk 4]
```diff
@@ -2,15 +2,15 @@
 apiVersion: v1
 kind: Namespace
 metadata:
-  name: test-kafka-strimzi
+  name: test-kafka-strimzi-cluster-provisioner-kafka
```
Comment (psav) [inline bundle/tests/scorecard/kuttl/test-kafka-strimzi-kafka-provisioner/00-install.yaml:5]: The name of the test and the name of the _root_ namespace do not match.

File: bundle/tests/scorecard/kuttl/test-kafka-strimzi-topic/00-install.yaml [hunk 5]
```diff
@@ -0,0 +1,41 @@
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: test-kafka-strimzi
```
Comment (psav) [inline bundle/tests/scorecard/kuttl/test-kafka-strimzi-topic/00-install.yaml:5]: The namespace here should match the test-name, or visa versa

File: controllers/cloud.redhat.com/providers/kafka/strimzi.go [hunk 6]
```diff
@@ -27,27 +28,191 @@ type strimziProvider struct {
...
+			Version:  &version,
+			Replicas: replicas,
+			Listeners: []strimzi.KafkaSpecKafkaListenersElem{
+				strimzi.KafkaSpecKafkaListenersElem{Name: "tcp", Type: "internal", Tls: false, Port: 9092},
+				strimzi.KafkaSpecKafkaListenersElem{Name: "tls", Type: "internal", Tls: true, Port: 9093},
+			},
+			Storage: strimzi.KafkaSpecKafkaStorage{
+				Type:        strimzi.KafkaSpecKafkaStorageTypePersistentClaim,
+				Size:        &storageSize,
+				DeleteClaim: &deleteClaim,
```
Comment (psav) [inline controllers/cloud.redhat.com/providers/kafka/strimzi.go:85]: We already have the `pvc` option for the providers, can we piggy back off this? Is it likely that you'll want zookeeper and kafka to have different PVC policies?

**Review by psav — CHANGES_REQUESTED** [review]
@BlakeHolifield and @psav took a look through - some changes need making here - there is also a bigger question about the API change - but that will have to wait till tomorrow.

File: build/minikube_setup.sh [hunk 7]
```diff
@@ -16,41 +16,61 @@ if ! command -v yq; then
...
 
 echo "*** Installing Strimzi resources ..."
 kubectl apply -f . -n $OPERATOR_NS
+cd -
+
+PROM_VERSION=0.45.0
+PROM_TARFILE="prometheus-operator-${PROM_VERSION}.tar.gz"
+if ! test -f ${PROM_TARFILE}; then
+    echo "*** Downloading ${PROM_TARFILE} ..."
+    curl -Ls -o ${PROM_TARFILE} https://github.com/prometheus-operator/prometheus-operator/archive/v${PROM_VERSION}.tar.gz
```
[the remark this answers is not in the document]
Comment (bsquizz) [reply build/minikube_setup.sh:61]: Are you ok if we download to a consistent folder name that is in `.gitignore`? The goal was to not have to re-download the .tar.gz on my local box every time minikube_setup.sh is run.

**Comment by bsquizz** [comment]
Thanks for the review @psav @BlakeHolifield  -- see https://github.com/RedHatInsights/clowder/pull/214/commits/6261743ad18068cf2b48a122a795feea3632d7f9 for all the incorporated changes. Also added one new commit to tweak minikube_setup.sh to only install operators if they are not present on the cluster already.

File: pr_check.sh [hunk 8]
```diff
@@ -28,4 +28,28 @@ export KUBEBUILDER_ASSETS=$PWD/kubebuilder_2.3.1_linux_amd64/bin
 IMG=$IMAGE_NAME:$IMAGE_TAG make docker-build
 IMG=$IMAGE_NAME:$IMAGE_TAG make docker-push
 
-docker run -i -v $PWD:/srcroot -e IMAGE_NAME=$IMAGE_NAME -e IMAGE_TAG=$IMAGE_TAG -e QUAY_USER=$QUAY_USER -e QUAY_TOKEN=$QUAY_TOKEN -e MINIKUBE_HOST=$MINIKUBE_HOST -e MINIKUBE_ROOTDIR=$MINIKUBE_ROOTDIR -e MINIKUBE_USER=$MINIKUBE_USER quay.io/psav/clowder_pr_check:blarg24 /srcroot/build/pr_check_inner.sh
+CONTAINER_NAME="clowder-pr-check-$ghprbPullId"
```
Comment (BlakeHolifield) [inline pr_check.sh:31]: `ghprbPullId` is a Jenkins var for the PR number, correct? So this would be `clowder-pr-check-214` ?

File: go.mod [hunk 9]
```diff
@@ -3,7 +3,7 @@ module cloud.redhat.com/clowder/v2
 go 1.13
 
 require (
-	github.com/RedHatInsights/strimzi-client-go v0.21.1-2
+	github.com/RedHatInsights/strimzi-client-go v0.21.1-3
```
Comment (BlakeHolifield) [inline go.mod:6]: Do we need this minor update or did this sneak. Go has some sneaky libraries :upside_down_face:

File: go.sum [hunk 10]
```diff
@@ -33,8 +33,8 @@ github.com/PuerkitoBio/purell v1.1.0/go.mod h1:c11w/QuzBsJSee3cPx9rAFu61PvFxuPbt
 github.com/PuerkitoBio/purell v1.1.1/go.mod h1:c11w/QuzBsJSee3cPx9rAFu61PvFxuPbtSwDGJws/X0=
 github.com/PuerkitoBio/urlesc v0.0.0-20160726150825-5bd2802263f2/go.mod h1:uGdkoq3SwY9Y+13GIhn11/XLaGBb4BfwItxLd5jeuXE=
 github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578/go.mod h1:uGdkoq3SwY9Y+13GIhn11/XLaGBb4BfwItxLd5jeuXE=
-github.com/RedHatInsights/strimzi-client-go v0.21.1-2 h1:2FV/RhXGbIqHzgWqTSmlw5jjSnTzR37Zi/yBX7KrYZg=
-github.com/RedHatInsights/strimzi-client-go v0.21.1-2/go.mod h1:DQ0Lv64xIvOwDNd+WQF7jWiF7G4pci2pKZsgDFk0k78=
+github.com/RedHatInsights/strimzi-client-go v0.21.1-3 h1:9McbENXUSe+r1bfFxjdsyAaJghwHvoZMYn1vn7HHm3g=
```
Comment (BlakeHolifield) [inline go.sum:36]: Same question about sneaky stuff as above

File: build/minikube_setup.sh [hunk 11]
```diff
@@ -15,42 +15,90 @@ if ! command -v yq; then
...
+fi
 
 echo "*** Waiting for Strimzi operator to come up ..."
-kubectl rollout status deployment/strimzi-cluster-operator -n $OPERATOR_NS
+kubectl rollout status deployment/strimzi-cluster-operator -n $STRIMZI_OPERATOR_NS
+
+echo "*** Waiting for Prometheus operator to come up ..."
+kubectl rollout status deployment/prometheus-operator -n $PROM_OPERATOR_NS
+
+cd -
```
Comment (BlakeHolifield) [inline build/minikube_setup.sh:104]: it _looks_ like this is jumping you back to the prommie bundle instead of the root.

**Review by BlakeHolifield — COMMENTED** [review]
Bit more feedback. Once the PR check is happy I think we're good to go. We can see if Kyle/Pete want to discuss the API stuff first or merge and then refactor after.

File: pr_check.sh (hunk 8 above)
Comment (bsquizz) [reply pr_check.sh:31]: Correct- they use the github pull request builder which provides this env var.

If you were to run this script locally, your container name would just end up being `clowder-pr-check-` which isn't great, but not terrible either.

File: go.mod (hunk 9 above)
Comment (bsquizz) [reply go.mod:6]: Needed for KafkaConnect

I am keeping the version name of his library (0.21.1) to match the version of Strimzi operator. I am just bumping the `-N` if I make changes to the package if its still based on Strimzi v0.21.1

File: build/minikube_setup.sh (hunk 11 above)
Comment (bsquizz) [reply build/minikube_setup.sh]: Yeah... ok, changed `cd $DOWNLOAD_DIR` and `cd -` to be only called in the `install` functions now so all the dir changes are "contained" within the function calls.

File: pr_check.sh (hunk 8 above)
Comment (BlakeHolifield) [reply pr_check.sh:31]: Okay gotcha. This was mostly for my education on what that was doing. Ty!

File: build/minikube_setup.sh (hunk 11 above)
Comment (bsquizz) [reply build/minikube_setup.sh]: annnnd bit by another 'cd' issue... fixed again. Just going to call `pwd` at the start of the script and keep cd'ing back to that dir at the end of those functions.

File: build/minikube_setup.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (kylape) [reply build/minikube_setup.sh]: sounds reasonable to me.  maybe have it be a dot folder

**Comment by bsquizz** [comment]
It's in `build/downloads` now -- which was added to .gitignore

File: build/minikube_setup.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (bsquizz) [reply build/minikube_setup.sh]: I opted for `build/downloads` -- which was added to .gitignore -- chose not to use a dot folder so it's not too sneaky.

[the remark this answers is not in the document]
Comment (bsquizz) [reply build/minikube_setup.sh]: Correction -- `build/operator_bundles`

File: build/minikube_setup.sh [hunk 12]
```diff
@@ -15,42 +15,91 @@ if ! command -v yq; then
...
+fi
+
+if [ -z "$PROM_DEPLOYMENT" ]; then
+    echo "*** Prometheus operator deployment not found, installing ..."
+    install_prometheus_operator
+else
+    echo "*** Prometheus operator deployment found, skipping install ..."
+fi
 
 echo "*** Waiting for Strimzi operator to come up ..."
```
Comment (kylape) [inline build/minikube_setup.sh:100]: could you background the `rollout status` commands in the respective `if` block and then call `wait` at the bottom instead?

File: build/minikube_setup.sh [hunk 13]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
```
Comment (kylape) [inline build/minikube_setup.sh:1]: the name of this script is fine :smile:

**Review by kylape — CHANGES_REQUESTED** [review]
I think the big question in my mind regarding this PR is: Do we want to continue to support using already-existing Kafka clusters in operator mode?  If not, then we can simplify a lot of the spec and implementation. If yes, then I might consider creating a two separate modes: operator and operator-no-cluster (or something like that).  This way we can leverage the mode option to simplify the other options in the spec.

File: build/minikube_setup.sh (hunk 13 above)
Comment (bsquizz) [reply build/minikube_setup.sh:1]: :tada:

**Comment by bsquizz** [comment]
> Do we want to continue to support using already-existing Kafka clusters in operator mode?

I think that mainly depends on if we want to eventually have Clowder creating KafkaTopic resources in stage and prod or if we intend to always control that via app-interface.

I'm not that satisfied with the way the spec looks either but I didn't want to break backward compatibility

Right now if you want "operator mode with no cluster provisioning" you have:

```
kafka:
  mode: operator
  namespace: kafka-ns
  clusterName: my-kafka-cluster
  connectNamespace: connect-ns
  connectClusterName: my-kafka-connect-cluster
```

If you want "operator mode with cluster provisioning" you have:
```
kafka:
  mode: operator
  namespace: kafka-ns
  clusterName: my-kafka-cluster
  connectNamespace: connect-ns
  connectClusterName: my-kafka-connect-cluster
  pvc: true
  kafkaClusterOptions:
    provisionCluster: true
    replicas: 1
  kafkaConnectClusterOptions:
    provisionCluster: true
    replicas: 1
```

The ideal "look and feel" of the spec, IMO, would be something like:
```
kafka:
  mode: operator
  kafka:
    namespace: kafka-ns
    name: my-kafka-cluster
    clusterProvisioner:
      enabled: true
      pvc: true
      replicas: 1
  kafkaConnect:
    namespace: connect-ns
    name: my-kafka-connect-cluster
    clusterProvisioner:
      enabled: true
      pvc: true
      replicas: 1
```

I think that spec is usable by all 3 modes (operator, local, and app-interface). We'd just leave out the "clusterProvisioner" entirely when not interested in having Clowder spin up the cluster. But obviously that will be a breaking change for any existing ClowdEnvironments we have floating around out there and especially for stage/prod we'd have to make sure we don't blow anything up

**Comment by bsquizz** [comment]
FYI we pulled the script enhancements out of this PR into https://github.com/RedHatInsights/clowder/pull/233

**Comment by kylape** [comment]
I like the idea spec much more, of course, and I think we should try to move to that if possible.  Would it be reasonable to provide backwards compat while moving to the idea spec?

Using the ideal spec as a starting point, I have a few questions/suggestions.

1. Do we need to have individual configuration for both kafka and kafka connect?  e.g. would you ever want PVCs for kafka but not kafka connect?
2. Can we assume that kafka connect and kafka will always live in the same namespace?  I know they technically can live in a different namespace, but are there ever situations we encounter where they actually do live in separate namespaces?
3. If we are going to deal with a breaking spec change here, should we try to future-proof a bit and consider how the spec will change when we add managed kafka as a mode?

**Comment by bsquizz** [comment]
OK, new spec implemented in https://github.com/RedHatInsights/clowder/pull/214/commits/29143e4a23c1a43db430de5b525bd8cd04df9f8e

File: controllers/cloud.redhat.com/providers/kafka/strimzi.go [hunk 14]
```diff
@@ -72,43 +269,48 @@ func NewStrimzi(p *p.Provider) (providers.ClowderProvider, error) {
 		},
 	}
 
+	// NOTE: we will not create ownerReferences on generated strimzi objects, since it is possible
+	// for them to reside a namespace that does not match the ClowdEnvironment's 'targetNamespace'
+	//
+	// Therefore, if a ClowdEnvironment is removed, we do not remove created strimzi objects
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/kafka/strimzi.go:275]: How do they get removed if not deleted along with an env?

File: controllers/cloud.redhat.com/providers/kafka/strimzi.go [hunk 15]
```diff
@@ -72,43 +269,48 @@ func NewStrimzi(p *p.Provider) (providers.ClowderProvider, error) {
 		},
 	}
 
+	// NOTE: we will not create ownerReferences on generated strimzi objects, since it is possible
+	// for them to reside a namespace that does not match the ClowdEnvironment's 'targetNamespace'
```
Comment (kylape) [inline controllers/cloud.redhat.com/providers/kafka/strimzi.go:273]: I didn't think this would be a reason to not set the owner ref

Comment (bsquizz) [reply controllers/cloud.redhat.com/providers/kafka/strimzi.go]: Set the owner ref as the ClowdEnvironment in https://github.com/RedHatInsights/clowder/pull/214/commits/ac9699b8ab0bc1459e395810731bf4e44d80a17f

File: bundle/tests/scorecard/kuttl/test-cyndi-strimzi/00-install.yaml [hunk 16]
```diff
@@ -13,3 +13,28 @@ metadata:
 spec:
   finalizers:
   - kubernetes
+---
```
[the remark this answers is not in the document]
Comment (bsquizz) [reply bundle/tests/scorecard/kuttl/test-cyndi-strimzi/00-install.yaml:16]: Good point. Yes, I should change these.

File: bundle/tests/scorecard/kuttl/test-kafka-strimzi-topic/00-install.yaml [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: test-kafka-strimzi-topic-kafka
+spec:
+  finalizers:
+  - kubernetes
+---
+# Install a kafka cluster so that the environment reconciler successfully finds 'listeners'
```
[the remark this answers is not in the document]
Comment (bsquizz) [reply bundle/tests/scorecard/kuttl/test-kafka-strimzi-topic/00-install.yaml]: A relic of the past when operator mode had optional kafka provisioning..

[the remark this answers is not in the document]
Comment (bsquizz) [reply bundle/tests/scorecard/kuttl/test-kafka-strimzi-topic/00-install.yaml]: (no longer needed)

**Review by psav — APPROVED** [review]

**Pull request merged**